type.md

```diff
@@ -250,13 +253,134 @@ The value of ARG is "Hello", which is not of type NUMBER.
 ...
 ~~~
 
-What about compile-time type checking?
 
-Well, you may provide type information for variables, function arguments etc
-etc via the macro [`declare`][declare]. However, similar to `:type` slot
+## Compile-time type checking
+
+You may provide type information for variables, function arguments
+etc via the macros [`declare`][declare] and [`declaim`][declaim].
+However, similar to the `:type` slot
 introduced in [CLOS section][clos], the effects of type declarations are
 undefined in Lisp standard and are implementation specific. So there is no
-guarantee that Lisp compiler will perform compile-time type checking.
+guarantee that the Lisp compiler will perform compile-time type checking.
+
+However, it is possible, and SBCL is an implementation that does
+thorough type checking.
+
+### Declaring the type of variables
+
+Use the macro [`declaim`][declaim].
+
+Let's declare that our global variable `*name*` is a string (you can
+type the following in any order in the REPL):
+
+~~~lisp
+(declaim (type (string) *name*))
+(defparameter *name* "book")
+~~~
+
+Now if we try to set it with a bad type, we get a `simple-type-error`:
+
+~~~lisp
+(setf *name* :me)
+Value of :ME in (THE STRING :ME) is :ME, not a STRING.
+   [Condition of type SIMPLE-TYPE-ERROR]
+~~~
+
+We can do the same with our custom types. Let's quickly declare the type `list-of-strings`:
+
+~~~lisp
+(defun list-of-strings-p (list)
```

Comment:
reflecting on this, this should be distinguished as a run time custom type checker, as opposed to a static type checker.  Since this code has to run every time the type check occurs.

Reply:
mmh suggestion on how to phrase this welcome.

Reply:
this code has to run every time the type check occurs, but it doesn't always occur at run time. When we use the declaim type declarations, it happens when we compile a function (thus acting as a static type checker).